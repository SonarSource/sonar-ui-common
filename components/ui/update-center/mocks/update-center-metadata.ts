/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { MetaDataInformation, MetaDataVersionInformation } from '../update-center-metadata';

export function mockMetaDataVersionInformation(
  overrides?: Partial<MetaDataVersionInformation>
): MetaDataVersionInformation {
  return {
    version: '5.13',
    date: '2019-05-31',
    compatibility: '6.7',
    archived: false,
    downloadURL: 'https://example.com/sonar-java-plugin-5.13.0.18197.jar',
    changeLogUrl: 'https://example.com/sonar-java-plugin/release',
    ...overrides
  };
}

export function mockMetaDataInformation(
  overrides?: Partial<MetaDataInformation>
): MetaDataInformation {
  return {
    name: 'SonarJava',
    key: 'java',
    isSonarSourceCommercial: true,
    organization: {
      name: 'SonarSource',
      url: 'http://www.sonarsource.com/'
    },
    category: 'Languages',
    license: 'SonarSource',
    issueTrackerURL: 'https://jira.sonarsource.com/browse/SONARJAVA',
    sourcesURL: 'https://github.com/SonarSource/sonar-java',
    versions: [
      mockMetaDataVersionInformation({ version: '2.0' }),
      mockMetaDataVersionInformation({ version: '1.0', archived: true })
    ],
    ...overrides
  };
}
