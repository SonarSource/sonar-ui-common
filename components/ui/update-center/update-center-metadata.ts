/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export interface MetaDataInformation {
  category?: string;
  isSonarSourceCommercial?: boolean;
  issueTrackerURL?: string;
  key?: string;
  license?: string;
  name: string;
  organization?: {
    name: string;
    url?: string;
  };
  sourcesURL?: string;
  versions?: MetaDataVersionInformation[];
}

export interface MetaDataVersionInformation {
  archived?: boolean;
  changeLogUrl?: string;
  compatibility?: string;
  date?: string;
  description?: string;
  downloadURL?: string | AdvancedDownloadUrl[];
  version: string;
}

export interface AdvancedDownloadUrl {
  label?: string;
  url: string;
}
