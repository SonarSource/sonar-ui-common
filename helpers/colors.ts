/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable no-bitwise, no-mixed-operators */
export function stringToColor(str: string) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  let color = '#';
  for (let i = 0; i < 3; i++) {
    const value = (hash >> (i * 8)) & 0xff;
    color += ('00' + value.toString(16)).substr(-2);
  }
  return color;
}

export function isDarkColor(color: string) {
  color = color.substr(1);
  if (color.length === 3) {
    // shortcut notation: #f90
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }
  const rgb = parseInt(color.substr(1), 16);
  const r = (rgb >> 16) & 0xff;
  const g = (rgb >> 8) & 0xff;
  const b = (rgb >> 0) & 0xff;
  const luma = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  return luma < 140;
}

export function getTextColor(background: string, dark = '#222', light = '#fff') {
  return isDarkColor(background) ? light : dark;
}
