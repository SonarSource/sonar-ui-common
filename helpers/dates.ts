/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as parse from 'date-fns/parse';

function pad(number: number) {
  if (number < 10) {
    return '0' + number.toString();
  }
  return number;
}

type ParsableDate = string | number | Date;

export function parseDate(rawDate: ParsableDate): Date {
  return parse(rawDate);
}

export function toShortNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function toNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return date.toISOString().replace(/\..+Z$/, '+0000');
}

export function isValidDate(date: Date): boolean {
  return !isNaN(date.getTime());
}
