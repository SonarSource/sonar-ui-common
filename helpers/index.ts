/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export * from './colors';
export * from './cookies';
export * from './csv';
export * from './dates';
export * from './getHistory';
export * from './handleRequiredAuthentication';
export * from './init';
export * from './keycodes';
export * from './l10n';
export * from './measures';
export * from './pages';
export * from './path';
export * from './query';
export * from './ratings';
export * from './request';
export * from './scrolling';
export * from './search';
export {
  get as getFromLocalStorage,
  remove as removeFromLocalStorage,
  save as saveInLocalStorage,
} from './storage';
export * from './strings';
export * from './testUtils';
export * from './urls';
