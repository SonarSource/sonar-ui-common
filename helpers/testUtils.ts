/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { ReactWrapper, ShallowWrapper } from 'enzyme';

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', mockEvent(event));
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export const KEYCODE_MAP: { [keycode: number]: string } = {
  13: 'enter',
  37: 'left',
  38: 'up',
  39: 'right',
  40: 'down'
};

export function keydown(key: number | string): void {
  let keyCode;
  if (typeof key === 'number') {
    keyCode = key;
  } else {
    // eslint-disable-next-line no-console
    console.warn('Using strings in keydown() is deprecated. Consider using the KeyCodes enum.');
    const mapped = Object.entries(KEYCODE_MAP).find(([_, value]) => value === key);
    if (!mapped) {
      throw new Error(`Cannot map key "${key}" to a keyCode!`);
    }
    keyCode = mapped[0];
  }

  const event = new KeyboardEvent('keydown', { keyCode, which: keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.documentElement.clientHeight/clientWidth` are getters by default,
  // so we need to redefine them. Pass `configurable: true` to allow to redefine
  // the properties multiple times.
  if (width) {
    Object.defineProperty(document.documentElement, 'clientWidth', {
      configurable: true,
      value: width
    });
  }
  if (height) {
    Object.defineProperty(document.documentElement, 'clientHeight', {
      configurable: true,
      value: height
    });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function scrollTo({ left = 0, top = 0 }) {
  Object.defineProperty(window, 'pageYOffset', { value: top });
  Object.defineProperty(window, 'pageXOffset', { value: left });
  const resizeEvent = new Event('scroll');
  window.dispatchEvent(resizeEvent);
}

export function setNodeRect({ width = 50, height = 50, left = 0, top = 0 }) {
  const { findDOMNode } = require('react-dom');
  const element = document.createElement('div');
  Object.defineProperty(element, 'getBoundingClientRect', {
    value: () => ({ width, height, left, top })
  });
  findDOMNode.mockReturnValue(element);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}
