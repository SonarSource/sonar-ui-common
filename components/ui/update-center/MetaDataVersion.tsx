/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import classNames from 'classnames';
import * as React from 'react';
import { AdvancedDownloadUrl, MetaDataVersionInformation } from './update-center-metadata';

export interface MetaDataVersionProps {
  versionInformation: MetaDataVersionInformation;
}

export default function MetaDataVersion(props: MetaDataVersionProps) {
  const {
    versionInformation: {
      archived,
      changeLogUrl,
      compatibility,
      date,
      description,
      downloadURL,
      version
    }
  } = props;

  const fallbackLabel = 'Download';

  const advancedDownloadUrls = isAdvancedDownloadUrlArray(downloadURL)
    ? downloadURL.map(url => ({ ...url, label: url.label || fallbackLabel }))
    : [{ label: fallbackLabel, url: downloadURL }];

  return (
    <div
      className={classNames('update-center-meta-data-version', {
        'update-center-meta-data-version-archived': archived
      })}>
      <div className="update-center-meta-data-version-version">{version}</div>

      <div className="update-center-meta-data-version-release-info">
        {date && <time className="update-center-meta-data-version-release-date">{date}</time>}

        {compatibility && (
          <span className="update-center-meta-data-version-compatibility">{compatibility}</span>
        )}
      </div>

      {description && (
        <div className="update-center-meta-data-version-release-description">{description}</div>
      )}

      {(advancedDownloadUrls.length > 0 || changeLogUrl) && (
        <div className="update-center-meta-data-version-release-links">
          {advancedDownloadUrls.length > 0 &&
            advancedDownloadUrls.map(
              (advancedDownloadUrl, i) =>
                advancedDownloadUrl.url && (
                  // eslint-disable-next-line react/no-array-index-key
                  <span className="update-center-meta-data-version-download" key={i}>
                    <a href={advancedDownloadUrl.url} rel="noopener noreferrer" target="_blank">
                      {advancedDownloadUrl.label}
                    </a>
                  </span>
                )
            )}

          {changeLogUrl && (
            <span className="update-center-meta-data-version-release-notes">
              <a href={changeLogUrl} rel="noopener noreferrer" target="_blank">
                Release notes
              </a>
            </span>
          )}
        </div>
      )}
    </div>
  );
}

function isAdvancedDownloadUrlArray(
  downloadUrl: string | AdvancedDownloadUrl[] | undefined
): downloadUrl is AdvancedDownloadUrl[] {
  return !!downloadUrl && typeof downloadUrl !== 'string';
}
