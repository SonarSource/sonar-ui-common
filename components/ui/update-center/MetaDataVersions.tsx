/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import * as React from 'react';
import MetaDataVersion from './MetaDataVersion';
import { MetaDataVersionInformation } from './update-center-metadata';

interface Props {
  versions: MetaDataVersionInformation[];
}

interface State {
  collapsed: boolean;
}

export default class MetaDataVersions extends React.Component<Props, State> {
  state: State = {
    collapsed: true,
  };

  componentDidUpdate(prevProps: Props) {
    if (prevProps.versions !== this.props.versions) {
      this.setState({ collapsed: true });
    }
  }

  handleClick = (event: React.SyntheticEvent<HTMLButtonElement>) => {
    event.preventDefault();
    event.currentTarget.blur();
    this.setState(({ collapsed }) => ({ collapsed: !collapsed }));
  };

  render() {
    const { versions } = this.props;
    const { collapsed } = this.state;

    const archivedVersions = versions.filter((version) => version.archived);
    const currentVersions = versions.filter((version) => !version.archived);

    return (
      <div className="update-center-meta-data-versions">
        {archivedVersions.length > 0 && (
          <button
            className="update-center-meta-data-versions-show-more"
            onClick={this.handleClick}
            type="button">
            {collapsed ? 'Show more versions' : 'Show fewer versions'}
          </button>
        )}

        {currentVersions.map((versionInformation) => (
          <MetaDataVersion
            key={versionInformation.version}
            versionInformation={versionInformation}
          />
        ))}

        {!collapsed &&
          archivedVersions.map((archivedVersionInformation) => (
            <MetaDataVersion
              key={archivedVersionInformation.version}
              versionInformation={archivedVersionInformation}
            />
          ))}
      </div>
    );
  }
}
