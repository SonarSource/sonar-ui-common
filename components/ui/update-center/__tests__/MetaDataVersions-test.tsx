/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import { click } from '../../../../helpers/testUtils';
import MetaDataVersion from '../MetaDataVersion';
import MetaDataVersions from '../MetaDataVersions';
import { mockMetaDataVersionInformation } from '../mocks/update-center-metadata';

it('should render correctly', () => {
  const wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();
});

it('should properly handle show more / show less', () => {
  const wrapper = shallowRender();
  expect(wrapper.find(MetaDataVersion).length).toBe(1);

  click(wrapper.find('.update-center-meta-data-versions-show-more'));
  expect(wrapper.find(MetaDataVersion).length).toBe(3);
});

function shallowRender(props?: Partial<MetaDataVersions['props']>) {
  return shallow<MetaDataVersions>(
    <MetaDataVersions
      versions={[
        mockMetaDataVersionInformation({ version: '3.0' }),
        mockMetaDataVersionInformation({ version: '2.0', archived: true }),
        mockMetaDataVersionInformation({ version: '1.0', archived: true })
      ]}
      {...props}
    />
  );
}
