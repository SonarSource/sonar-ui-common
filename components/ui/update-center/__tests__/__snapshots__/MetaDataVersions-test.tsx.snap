// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="update-center-meta-data-versions"
>
  <button
    className="update-center-meta-data-versions-show-more"
    onClick={[Function]}
    type="button"
  >
    Show more versions
  </button>
  <MetaDataVersion
    key="3.0"
    versionInformation={
      Object {
        "archived": false,
        "changeLogUrl": "https://example.com/sonar-java-plugin/release",
        "compatibility": "6.7",
        "date": "2019-05-31",
        "downloadURL": "https://example.com/sonar-java-plugin-5.13.0.18197.jar",
        "version": "3.0",
      }
    }
  />
</div>
`;
